File: src/auth/device-login.ts
import { hostname } from 'os';
import { CliError } from '../utils/errors';
import type { ApiKeyView } from './api-keys';

/**
 * Client side of the hub's device login (`src/daemon/device-auth.ts`): ask
 * for a code, show it, poll until the owner decides. No callback server and
 * no redirect, so it works over SSH and in containers; the token only ever
 * travels in the poll answer.
 */

export interface DeviceLoginOptions {
  /** Hub base URL as the user typed it; only its origin is used. */
  url: string;
  /** Shown to the owner and used as the key's default name. */
  name?: string;
  /** Called once with the code and the page the owner must open. */
  onCode: (info: { userCode: string; verifyUrl: string; expiresIn: number }) => void;
  /** Tests override the hub's interval; production honours what the hub says. */
  pollMs?: number;
}

export interface DeviceLoginResult {
  url: string;
  token: string;
  key: ApiKeyView;
}

const REQUEST_TIMEOUT_MS = 15_000;

/** Origin of the hub URL, or INVALID_PARAMS. */
export function hubOrigin(input: string): string {
  let parsed: URL;
  try {
    parsed = new URL(/^[a-z][a-z0-9+.-]*:\/\//i.test(input) ? input : `https://${input}`);
  } catch {
    throw new CliError('INVALID_PARAMS', `Not a valid hub URL: ${input}`, 'Pass the hub as https://vault.example.com');
  }
  if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') {
    throw new CliError('INVALID_PARAMS', 'The hub URL must use http or https');
  }
  return parsed.origin;
}

async function post<T>(url: string, path: string, body: unknown): Promise<T> {
  let response: Response;
  try {
    response = await fetch(`${url}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err);
    throw new CliError('NETWORK_ERROR', `Cannot reach the vault hub at ${url}: ${reason}`, 'Check the URL, the network, and that the hub daemon is running');
  }
  const text = await response.text();
  let parsed: { error?: string; suggestion?: string } | null = null;
  try { parsed = text ? JSON.parse(text) : null; } catch { parsed = null; }
  if (!response.ok) {
    if (response.status === 404 && path === '/v1/device') {
      throw new CliError('CONFIG_ERROR', `${url} does not offer device login`, 'Is this the hub URL, and is the hub up to date?');
    }
    throw new CliError(response.status === 429 ? 'RATE_LIMITED' : 'API_ERROR', parsed?.error ?? `HTTP ${response.status}`, parsed?.suggestion);
  }
  return parsed as T;
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

type Poll =
  | { status: 'pending'; interval: number }
  | { status: 'denied' }
  | { status: 'approved'; token: string; key: ApiKeyView };

export async function deviceLogin(options: DeviceLoginOptions): Promise<DeviceLoginResult> {
  const url = hubOrigin(options.url);
  const name = options.name?.trim() || hostname();
  const start = await post<{ userCode: string; deviceCode: string; expiresIn: number; interval: number }>(url, '/v1/device', { name });
  options.onCode({ userCode: start.userCode, verifyUrl: `${url}/ui#authorize=${start.userCode}`, expiresIn: start.expiresIn });

  const deadline = Date.now() + start.expiresIn * 1000;
  let interval = options.pollMs ?? start.interval * 1000;
  while (Date.now() < deadline) {
    await sleep(interval);
    let answer: Poll;
    try {
      answer = await post<Poll>(url, '/v1/device/token', { deviceCode: start.deviceCode });
    } catch (err) {
      // 404 is the hub saying the request is gone: expired, or it restarted.
      if (err instanceof CliError && err.code === 'API_ERROR' && /expired/i.test(err.message)) break;
      if (err instanceof CliError && err.code === 'RATE_LIMITED') { interval *= 2; continue; }
      throw err;
    }
    if (answer.status === 'approved') return { url, token: answer.token, key: answer.key };
    if (answer.status === 'denied') throw new CliError('AUTH_FAILED', 'The hub owner denied this login');
    if (options.pollMs === undefined) interval = answer.interval * 1000;
  }
  throw new CliError('AUTH_FAILED', 'The login code expired before it was approved', 'Run `agentio login` again and approve within ten minutes');
}

File: go/internal/profile/device_login.go
package profile

import (
	"encoding/json"
	"errors"
	"math"
	"net/http"
	"os"
	"regexp"
	"time"

	"github.com/plosson/agentio/go/internal/auth"
	"github.com/plosson/agentio/go/internal/clierr"
	"github.com/plosson/agentio/go/internal/jsvalue"
)

// Client side of the hub's device login (Bun src/auth/device-login.ts): ask
// for a code, show it, poll until the owner decides.

type DeviceLoginOptions struct {
	// URL is the hub as typed; only its origin is used, https when no scheme.
	URL string
	// Name introduces this machine; blank is the hostname.
	Name string
	// OnCode is called once with the code, the page the owner must open and
	// the hub's expiresIn (seconds).
	OnCode func(userCode, verifyURL string, expiresIn float64)
	// PollInterval replaces the interval the hub asks for (tests).
	PollInterval time.Duration
	// Sleep waits between polls; nil is time.Sleep.
	Sleep func(time.Duration)
}

type DeviceLoginResult struct {
	URL   string
	Token string
	Key   KeyView
}

var schemePrefix = regexp.MustCompile(`(?i)^[a-z][a-z0-9+.-]*://`)

// HubOrigin is the origin of the hub URL, https assumed without a scheme.
func HubOrigin(input string) (string, error) {
	if !schemePrefix.MatchString(input) {
		input = "https://" + input
	}
	return ValidateHubURL(input)
}

func errorCode(err error) clierr.Code {
	var ce *clierr.Error
	if errors.As(err, &ce) {
		return ce.Code
	}
	return ""
}

// jsNumber is `value * 1` for a JSON value: NaN for anything that is not a
// number, a numeric string, a boolean or null.
func jsNumber(v any) float64 {
	switch t := v.(type) {
	case nil:
		return 0
	case float64:
		return t
	case bool:
		if t {
			return 1
		}
		return 0
	case string:
		return jsvalue.Number(t)
	}
	return math.NaN()
}

// timeoutDelay is setTimeout's delay: 1 ms unless it is a number from 1 to 2^31-1.
func timeoutDelay(ms float64) time.Duration {
	if !(ms >= 1 && ms <= math.MaxInt32) {
		ms = 1
	}
	return time.Duration(ms * float64(time.Millisecond))
}

func DeviceLogin(opts DeviceLoginOptions) (DeviceLoginResult, error) {
	origin, err := HubOrigin(opts.URL)
	if err != nil {
		return DeviceLoginResult{}, err
	}
	name := jsvalue.Trim(opts.Name)
	if name == "" {
		name, _ = os.Hostname()
	}
	sleep := opts.Sleep
	if sleep == nil {
		sleep = time.Sleep
	}
	notAHub := clierr.New(clierr.ConfigError, origin+" does not offer device login", "Is this the hub URL, and is the hub up to date?")
	raw, _, err := auth.HubCall(origin, "/v1/device", auth.Call{Method: http.MethodPost, Body: map[string]string{"name": name}})
	if err != nil {
		// Not a hub at all (404), or a hub too old to have the route, which
		// answers with its bearer check instead.
		if code := errorCode(err); code == clierr.NotFound || code == clierr.AuthFailed {
			return DeviceLoginResult{}, notAHub
		}
		return DeviceLoginResult{}, err
	}
	// A landing page answers 200 with HTML, which is no JSON object.
	var start map[string]any
	if json.Unmarshal(raw, &start) != nil || start == nil {
		return DeviceLoginResult{}, notAHub
	}
	userCode, ok1 := start["userCode"].(string)
	deviceCode, ok2 := start["deviceCode"].(string)
	if !ok1 || !ok2 {
		return DeviceLoginResult{}, notAHub
	}
	expiresIn := math.NaN()
	if v, ok := start["expiresIn"]; ok {
		expiresIn = jsNumber(v)
	}
	if opts.OnCode != nil {
		opts.OnCode(userCode, origin+"/ui#authorize="+userCode, expiresIn)
	}
	nowMs := func() float64 { return float64(time.Now().UnixMilli()) }
	deadline := nowMs() + expiresIn*1000
	interval := float64(opts.PollInterval) / float64(time.Millisecond)
	if opts.PollInterval == 0 {
		interval = math.NaN()
		if v, ok := start["interval"]; ok {
			interval = jsNumber(v) * 1000
		}
	}
	for nowMs() < deadline {
		sleep(timeoutDelay(interval))
		body, _, err := auth.HubCall(origin, "/v1/device/token", auth.Call{Method: http.MethodPost, Body: map[string]string{"deviceCode": deviceCode}})
		if err != nil {
			code := errorCode(err)
			if code == clierr.NotFound {
				break // the hub forgot it: expired, or restarted
			}
			if code == clierr.RateLimited {
				interval *= 2
				continue
			}
			return DeviceLoginResult{}, err
		}
		var answer struct {
			Status string  `json:"status"`
			Token  string  `json:"token"`
			Key    KeyView `json:"key"`
		}
		_ = json.Unmarshal(body, &answer)
		switch answer.Status {
		case "approved":
			return DeviceLoginResult{URL: origin, Token: answer.Token, Key: answer.Key}, nil
		case "denied":
			return DeviceLoginResult{}, clierr.New(clierr.AuthFailed, "The hub owner denied this login", "")
		}
	}
	return DeviceLoginResult{}, clierr.New(clierr.AuthFailed, "The login code expired before it was approved",
		"Run `agentio login` again and approve within ten minutes")
}
